### DISCUSSION TITLE
Attribute to Change Test Result or Retry Test Based on Error

### DISCUSSION THREAD

**Discussion by stevemerckel** [nunit/nunit#3975] [General]
I have some integration tests that bump against an API.  Occasionally, the API will cycle and return a HTTP 504 response.  My question is whether this specific response can be intercepted to force a retry attempt (similar to the RetryAttribute behavior), or if the more pragmatic approach is to change the test result from Error to Inconclusive.

I took a shot at writing up the latter idea, but it is incomplete at the "do... something" code comment:
```C#
/// <summary>
/// Decorates a test to indicate that HTTP 504 responses (Gateway Timeout) should be treated as "Inconclusive"
/// </summary>
[AttributeUsage(AttributeTargets.Class|AttributeTargets.Method|AttributeTargets.Assembly, AllowMultiple=true, Inherited=true)]
public sealed class GatewayTimeoutInconclusiveAttribute : NUnitAttribute, IWrapSetUpTearDown
{
    /// <inheritdoc />
    public TestCommand Wrap(TestCommand command)
    {
        var testResult = command.Test.MakeTestResult();
        if (testResult.ResultState == ResultState.Error || testResult.ResultState == ResultState.Failure)
        {
            var message = testResult.Message;
            if (!string.IsNullOrWhiteSpace(message))
            {
                if (message.Contains("504") || message.Contains("Gateway Timeout", StringComparison.InvariantCultureIgnoreCase))
                {
                    // do... something
                }
            }
        }

        return command;
    }
}
```

My hope is to have this as a block of consolidated logic.  I previously tried a method decorated with "TearDown", but did not see where to change the result of the test context -- adding "Assert.Inconclusive" was appended to the existing "Error" originating from the test.  Looking through the docs pointed me to Execution-time attributes, and I tried the idea shown above.

Is something like the above code sample possible, or even a pragmatic approach to my problem?  Am I re-inventing the wheel of another feature that does this work?

Thanks!

**Comment by stevenaw** [comment]
Hi @stevemerckel 
This is a great question. NUnit does define a "Retry" attribute, but it doesn't support the fine-grained retry conditions you're seeking (it's all-or-nothing). The attribute uses the `IRepeatTest` interface to define the retry hook, have you tried adding that?
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Attributes/RetryAttribute.cs#L17

The "RetryCommand" which does most of this work should be a public class, so it may be as simple for you as defining your own `Wrap` method which returns a RetryCommand:
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Attributes/RetryAttribute.cs#L37-L40

Another option if you wanted the retry to occur within your code itself could be to look into using a third-party http fault tolerance library  like Polly: https://github.com/App-vNext/Polly

**Reply by stevemerckel** [reply]
Thank you for the thorough response, @stevenaw !!

The `IRepeatTest` hook is a good idea, I'll try that out soon.

And I do like your other suggestion for _Polly_.  I can see that as providing a significant amount of value during unpredictable periods for the server while the app runtime is working -- which would (transparently) benefit the integration tests, too.
